---
title: "TRGN510_Final_Project"
author: "Mengyuan Zhang"
date: "10/25/2020"
output:
  html_document: default
  pdf_document: default
---

# Load in packages that will be used in this project
```{r}
library(stringr)
library(dplyr)
library(Homo.sapiens)
library(gsubfn)
library(limma)
library(edgeR)
```

```{r}
options(stringsAsFactors = FALSE)
```

# Data Packaging
## Check current working directory
```{r}
getwd()
```

## Set working directory
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project/raw_young")
```

## Load and integrate the data of young group
Create a new folder named "row_young_data" to store the data of young group. 
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project")
dir.create('raw_young_data')
```

Find all of files contained in the original data folder "raw_young", and copy all of the files into the folder "row_young_data".
```{r}
for (dirname in dir('raw_young')){
  file <- list.files(paste0(getwd(),'/raw_young/',dirname), pattern = '*.counts')
  file.copy(paste0(getwd(),'/raw_young/',dirname,'/',file), 'raw_young_data')
}
```

Because the original names of files are so complicated, I am going to give these files new names for this project to easily indentify them. 
Create "list1" that containd original names of 20 files.
```{r}
list1 <- list.files("/Users/zmyyy/Desktop/510/final_project/raw_young_data", pattern = '*.gz')
list1
```
 
Create "list2" that containd the 20 new names from "young_1" to "young_20"
```{r}
list2 <- str_c(c("young"), c(1:20), sep = "_")
list2
```

Combine the original names and the new names of young group into one data frame named "young_name" , rename the column names of "young_data" as "file_names" and "project_names" to identify the original names and the new names
```{r}
young_name <- as.data.frame(list(list1,list2))
colnames(young_name) <- c("file_names", "project_names")
```

Read the first file of young group as a test to check the format
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project/raw_young_data")
test_young <- data.table::fread(young_name$file_names[1])
```

Create a data frame "young_data" containing all data of the 20 files of young group
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project/raw_young_data")
young_data <- data.frame(matrix(NA,nrow(test_young),nrow(young_name)))
for ( i in 1:nrow(young_name)) {
  print(i)
  young_data[,i] = data.table::fread(young_name$file_names[i])[,2]
}
colnames(young_data) <- young_name$project_names
```

## Load and integrate data of the old group
Do the same steps with the young group
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project")
dir.create('raw_old_data')
```

```{r}
for (dirname in dir('raw_old')){
  file <- list.files(paste0(getwd(),'/raw_old/',dirname), pattern = '*.counts')
  file.copy(paste0(getwd(),'/raw_old/',dirname,'/',file), 'raw_old_data')
}
```

```{r}
list3 <- list.files("/Users/zmyyy/Desktop/510/final_project/raw_old_data",pattern = '*.counts')
list3
```

```{r}
list4 <- str_c(c("old"), c(1:20), sep = "_")
list4
```

```{r}
old_name <- as.data.frame(list(list3,list4))
colnames(old_name) <- c("file_names", "project_names")
```


```{r}
setwd("/Users/zmyyy/Desktop/510/final_project/raw_old_data")
test_old <- data.table::fread(old_name$file_names[1])
old_data <- data.frame(matrix(NA,nrow(test_old),nrow(old_name)))
for ( i in 1:nrow(old_name)) {
  print(i)
  old_data[,i] = data.table::fread(old_name$file_names[i])[,2]
}
colnames(old_data) <- old_name$project_names
```

## Organize the data
We got two data frames, one contains the all data of young group, another contains th all data of old group. We need to combine these two data frames into one named "project_data".
```{r}
project_data <- cbind(young_data,old_data)
```

Create a data frame named "ensembl_id"containing all ensembl IDs
```{r}
setwd("/Users/zmyyy/Desktop/510/final_project/raw_old_data")
ensembl_id <- data.table::fread(old_name$file_names[i])$V1
```

Combine the "project_data" and "ensembl_id"
```{r}
project_data <- cbind(ensembl_id = ensembl_id, project_data)
```

Check the data of the last 6 rows 
```{r}
tail(project_data)
```

Data of the last 5 rows are different from the others, so we need to delete the last 5 rows data
```{r}
project_data <- project_data[1:(nrow(young_data)-5),]
```

Check if we successfully delete the last 5 rows
```{r}
tail(project_data)
```


The format of "ensemnl_id" in "project_data contains" "." and numbers, and doesn't macth the ENSEMBL IDs in library "Homo.sapiens". So the "ensemble_id" in "project_data" should be converted by using package "gsubfn".
```{r}
ENSEMBL<- gsub("\\.\\d*", "", project_data$ensembl_id)
head(ENSEMBL)
```

Create a dataframe named "genes" to store information of gene level. 
Retrive GeneIDs, "SYMBOL" and chromosome "TXCHROM" from "Homo.sapiens."
```{r}
genes <- select(Homo.sapiens, keys=ENSEMBL, columns=c("SYMBOL", "TXCHROM"), 
                keytype="ENSEMBL")
```

Delete the duplicated gene names.
```{r}
genes <- genes[!duplicated(genes$ENSEMBL),]
```

Due to the following steps will use function DGElist, it will read data but not ensemble IDs. To maintain the ensemble IDs in the data frame, I'm going to use the enseble IDs as the row names of "project_data" and remove the column of ensemble IDs from "project_data."
```{r}
row.names(project_data) <- c(project_data$ensembl_id)
project_data <- project_data[2:41]
```

Group data according to their age
```{r}
group <- c(rep("young",20), rep("old",20))
```

Create a datafram named "y", and it will contain the information of counts, group and genes.
```{r}
y <- DGEList(counts = project_data, group = group, genes = genes)
y
```

# Data Pre-processing
## Transformations from the raw-scale
The data is converted from the original format to a more useful format. For this purpose, calculations of "cpm: counts per million" and "lcpm: logarithms per million" are performed.
```{r}
cpm <- cpm(y)
lcpm <- cpm(y, log=TRUE)
```

For further calculations, the following calculations will be used to process the data. L is the mean library size, in millions. M is the median of the library, in millions.
```{r}
L <- mean(y$samples$lib.size) * 1e-6
M <- median(y$samples$lib.size) * 1e-6
c(L, M)
```

Present the summary of lcmp calculation
```{r}
summary(lcpm)
```

# Removing genes that are lowly expressed
Some genes have low expression and are meaningless, so these genes should be removed.
```{r}
table(rowSums(y$counts==0)==9)
```

There is another effecient way to filter genes with low expression.
```{r}
keep.exprs <- filterByExpr(y, group=group)
y <- y[keep.exprs,, keep.lib.sizes=FALSE]
dim(y)
```

After flitering, the genes reduced dramatically. Using the package RColorBrewer to generate graphs to show the raw data and flitered data. 
```{r}
lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(y)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
den <- density(lcpm[,i])
lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", group, text.col=col, bty="n")
lcpm <- cpm(y, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
den <- density(lcpm[,i])
lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", group, text.col=col, bty="n")
```


## Normalising gene expression distributions
The normalization of data is necessary for analysis. The resultant values can be stored as a norm.factor dataset.
```{r}
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors
```

To have a better understanding of data normalization, the counts of the young_1 were reduced to 1% and the counts of the young_2 were increased to 1000%
```{r}
y2 <- y
y2$samples$norm.factors <- 1
y2$counts[,1] <- ceiling(y2$counts[,1]*0.01)
y2$counts[,2] <- y2$counts[,2]*10
```

This is the graph using unnormalised data
```{r}
par(mfrow=c(1,2))
lcpm <- cpm(y2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
y2 <- calcNormFactors(y2)  
y2$samples$norm.factors
```

This is the graph using normalised data
```{r}
lcpm <- cpm(y2, log=TRUE)
boxplot(lcpm, las=2, col=col, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")
```


## Unsupervised clustering of data
Generate graphs by using plotMDS in limma, and the graphs will show the similarity and dissimilarity in samples.
```{r}
lcpm <- cpm(y, log=TRUE)
par(mfrow=c(1,1))
col.group <- group
col.group
levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
col.group <- as.character(col.group)
col.group <- c("purple","red")[group]
plotMDS(lcpm, labels=group, col=col.group)
title(main="A. Sample groups")
```

An interactive version of this graph can be launched using glimma.
```{r}
library(Glimma)
glMDSPlot(lcpm,groups = group)
```

# Differential expression analysis
## Creating a design matrix and contrasts
Create a design matrix with both the groups of data.
```{r}
design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))
design
```

Contrasts for the two groups are created and stored in a matrix
```{r}
contr.matrix <- makeContrasts(
   youngvsold = young-old, 
   levels = colnames(design))
contr.matrix
```

## Removing heteroscedascity from count data
Transform the raw data into the normalised data by using function "voom" in limma
```{r}
par(mfrow=c(1,2))
v <- voom(y, design, plot=TRUE)
v
```


```{r}
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")
```

## Examining the number of DE genes 
Use T-statistics to get significant differential gene expression
```{r}
summary(decideTests(efit))
```

Due to there is no differential gene expression was found from the last step, I changed "lfc" from 1 to 0.1 and "p.value" from 0.05 to 0.145 to find more possible significant DGE.
```{r}
tfit <- treat(vfit, lfc=0.1)
dt <- decideTests(tfit, p.value = 0.145)
summary(dt)
```

After the changes of "lfc" and "p.value", I get eight DGEs. 
```{r}
de.common <- which(dt[,1]!=0)
length(de.common)
```

```{r}
head(tfit$genes$SYMBOL[de.common], n=20)
```

```{r}
vennDiagram(dt[,1], circle.col=c("turquoise", "salmon"))
```

```{r}
write.fit(tfit, dt, file="results.txt")
```

## Examining individual DE genes from top to bottom
Use a function called "toptreat" to list the genes with the most expression differences.
```{r}
Young.vs.Old <- topTreat(tfit, coef=1, n=Inf)
head(Young.vs.Old)
```

## Useful graphical representations of differential expression results
To summarize genes, plotMD function is plotted with respect to the change of the number of counts the number of parts per million.
```{r}
plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], 
       xlim=c(-8,13))
```

Create an interactive graph by using Glimma to show the same information
```{r}
glMDPlot(tfit, coef=1, status=dt, main=colnames(tfit)[1],
         side.main="ENSEMBL", counts=lcpm, groups=group, launch=TRUE)
```

Plot the differential gene expressions in a heatmap
```{r}
library(gplots)
library(heatmap.plus)
Young.vs.Old.topgenes <- Young.vs.Old$ENSEMBL[1:50]
i <- which(v$genes$ENSEMBL %in% Young.vs.Old.topgenes)
mycol <- colorpanel(1000,"deepskyblue3","snow","firebrick3")
par(cex.main=0.8,mar=c(1,1,1,1))
heatmap.plus(lcpm[i,],  scale="row", labRow=v$genes$SYMBOL[i], labCol=group, col=mycol, cexRow=1,cexCol=0.2, margins = c(8,6), main="HeatMap")
```

Highly expressed genes are represented by red, lowly expressed genes are represented by blue, and intermediately expressed gene are represented by white. 



